docs/specification/2025-03-26/basic/authorization.mdx

```diff
@@ -63,248 +155,342 @@ audience. For instance:
+[Client Metadata](https://datatracker.ietf.org/doc/html/rfc7591#section-2) section of
+[RFC 7591](https://datatracker.ietf.org/doc/html/rfc7591).
+
+MCP clients with mass audiences will *generally* want to pre-register and hardcode client IDs.
+
+Some authorization servers (such as Google and Microsoft) limit the functionality of dynamically
+registered clients.
+
+An MCP client **MAY** contain a configuration for common issuers, such as:
+
+```json
+[
+  { "issuer":"https://authserver.example", "client_id":"12345"},
+  { "issuer":"https://authserver2.example", "client_id":"urn:client_ids:546"}
+]
+```
 
-1. Hardcode a client ID (and, if applicable, client secret) specifically for that MCP
-   server, or
-2. Present a UI to users that allows them to enter these details, after registering an
-   OAuth client themselves (e.g., through a configuration interface hosted by the
-   server).
+Alternatively, an MCP client could allow the user to register the MCP client and provide those details
+to the MCP client through a UI.
 
-### 2.5 Authorization Flow Steps
+### 2.8 Authorization flow steps
 
-The complete Authorization flow proceeds as follows:
+The complete authorization flow proceeds as follows:  
 
 ```mermaid
 sequenceDiagram
-    participant B as User-Agent (Browser)
-    participant C as Client
-    participant M as MCP Server
-
-    C->>M: GET /.well-known/oauth-authorization-server
-    alt Server Supports Discovery
-        M->>C: Authorization Server Metadata
-    else No Discovery
-        M->>C: 404 (Use default endpoints)
+    participant UA as User-Agent (Browser or App/Broker)
+    participant MC as MCP Client
+    participant MS as MCP Server
+    participant AS as Authorization Server
+
+    MC->>MS: Request data
+    MS-->>MC: 401 Unauthorized with WWW-Authenticate
+
+    MC->>MS: Fetch PRM document
+    MS-->>MC: PRM document response
+    Note over MC: Validate PRM document<br/>metadata
+    Note over MC: Build AS metadata URL
+    MC->>AS: Fetch AS metadata
+    AS-->>MC: AS metadata response
+
+    alt External authorization server
+        MC->>AS: GET /.well-known/oauth-authorization-server
+        AS-->>MC: 200 OK + Metadata Document
+        Note over MC: Use endpoints from metadata
+    else Co-located authorization server
+        MC->>MS: GET /.well-known/oauth-authorization-server
+        MS-->>MC: 200 OK + Metadata Document
+        Note over MC: Use endpoints from metadata
     end
 
-    alt Dynamic Client Registration
-        C->>M: POST /register
-        M->>C: Client Credentials
+    alt Client is pre-registered with the authorization server
+    else Authorization server supports dynamic client registration
+        MC->>AS: Request client registration
+        AS-->>MC: Confirm client registration
+    else Authorization server does not support dynamic client registration
+        MC->>MS: Request client registration
+        MS->>AS: Register client
```

Comment:
This step confuses me - the upstream AS does not supports DCR but the MS does? Wouldn't this make the MS the AS under the delegated auth case?

I think this section should instead show the `Alternatively, an MCP client could allow the user to register the MCP client and provide those details to the MCP client through a UI.` state described on line 359
```
MC->> UA: Request Out-of-band Client Registration
UA->>UA: Complete Out-of-band Client Registration
UA-->>MC: Return client registration 
```

Reply:
In this case, the MS could request client registration with the AS through exposed APIs, even if DCR is not supported directly (e.g., through AS-specific endpoints, like [Graph for Entra ID](https://learn.microsoft.com/en-us/graph/api/application-post-applications?view=graph-rest-1.0&tabs=http)). This is something I've also discussed with @allenzhou101 before in the context of Descope, where DCR might not be supported directly, but the workaround is to wrap AS registration APIs in a custom endpoint.

But you're right @max-stytch that at that point you become the AS (even if you redirect parts to your own, parts to third-party AS).